Move the metadata update related APIs to the MetadataUpdater class

mikem8361 opened the pull request:
The old APIs AssemblyExtensions.ApplyUpdate() and AssemblyExtensions.GetApplyUpdateCapabilities() will be removed
after all the references to them have been changed to the new APIs.

Add the new IsSupported API described in issue https://github.com/dotnet/runtime/issues/51159.

Change the tests to use the MetadataUpdater APIs.

/cc: @pranavkm, @eerhardt

Commit:
Move the metadata update related APIs to the MetadataUpdater class

The old APIs AssemblyExtensions.ApplyUpdate() and AssemblyExtensions.GetApplyUpdateCapabilities() will be removed
after all the references to them have been changed to the new APIs.

Add the new IsSupported API described in issue https://github.com/dotnet/runtime/issues/51159.

Change the tests to use the MetadataUpdater APIs.

mikem8361 commented:
@lambdageek, I'm sure the mono version of MetadataUpdater.IsSupported isn't complete, but there is a place holder.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        /// Returns true if the apply assembly update is enabled and available.
+        /// </summary>
+        public static bool IsSupported => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "" || Environment.GetEnvironmentVariable("COMPlus_ForceEnc") == "1";
```
It is not unusual for the user code to clear env variables like this so that they are not inherited by child processes; or on the other hand to set these env variable so that the child processes have them.

Should this rather return the setting cached by the runtime?

jkotas replied:
`Debugger.IsAttached` can change while the process is running. I do not think it is desirable behavior - it is hard to write reliable code when the `IsSupported` can be changing underneath you.

stephentoub replied:
> Should this rather return the setting cached by the runtime?

At a minimum we should be caching this so that the value doesn't change, e.g.
```suggestion
        public static bool IsSupported { get; } = Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "" || Environment.GetEnvironmentVariable("COMPlus_ForceEnc") == "1";
```

But, yes, it'd be better if we could ask the runtime for its view on whether updates could potentially be applied.  I'd suggested that previously and was told there was some reason we couldn't, but I don't remember what it was.

mikem8361 replied:
Yes, I forgot about caching this value like we talked about.  Hopefully what Stephen suggests is enough. The runtime really can't do any better via a QCall other than reading these as config values which can't be changed by the app.

jkotas replied:
If this reads the env variables independently on the runtime, this property can be out of sync with what runtime thinks. I think it would be better to just return the value that the runtime cached.

mikem8361 replied:
This roughly what I think the QCall would look like to get the runtime apply update supported state. It would still be cached in MetadataUpdater.IsSupported. Is this what you are thinking?
```
BOOL QCALLTYPE AssemblyNative::IsApplyUpdateSupported()
{
    QCALL_CONTRACT;

    BEGIN_QCALL;

#ifdef EnC_SUPPORTED
    BOOL result = CORDebuggerAttached() || g_pConfig->ForceEnc() || g_pConfig->DebugAssembliesModifiable();
#else
    BOOL result = false;
#endif

    END_QCALL;

    return result
}
```

jkotas replied:
Yes, something like this. Why do we need `CORDebuggerAttached` in the condition? I do not think attaching a debugger will turn on EnC today unless it was turned on earlier.

mikem8361 replied:
It is for the debugger launch case which most debuggers enable ENC for user modules.  This is a case where ENC can be enabled but neither env var has been set.  There really isn't any global state, but per module so CORDebuggerAttached is the closest thing for the launch case.

For debugger attach, either ForceEnc or DebugAssembliesModifiable configs would have to be set to enable ENC.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    public static class MetadataUpdater
+    {
+        [DllImport(RuntimeHelpers.QCall)]
```
This needs matching update in the table of QCalls.

jkotas commented on src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:
```diff
@@ -12,6 +12,12 @@ public static partial class AssemblyExtensions
     }
+    public static partial class MetadataUpdater
+    {
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta) { throw null; }
```
My original reason for having this API on AssemblyExtensions was to use it to implement modern equivalent of Reflection.Emit eventually. This new API shape does not make sense for the eventual Reflection.Emit use case.
Reflection.Emit assemblies would always support updating, independent on whether the debugger is attached, etc.

Just pointing it out. I can live with this shape. If/once we do something to modernize Reflection.Emit, we can introduce a different ApplyUpdate method in the right place.

mikem8361 replied:
The API review committee ask for the move to MetadataUpdater class:  https://github.com/dotnet/runtime/issues/51159#issuecomment-866199485

marek-safar commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,65 @@
+            return ApplyUpdateEnabled() != 0 ? "Baseline" : string.Empty ;
+        }
+
+        public static bool IsSupported => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "";
```
@lambdageek we want to introduce a feature switch backed up by msbuild property to use it to control build dependencies, right?

lambdageek replied:
Yea, we need a linker substitution for this flag, and also a runtime check

lambdageek replied:
```suggestion
        public static bool IsSupported => (Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "") && ApplyUpdateEnabled();
```

mikem8361 replied:
We may need a new icall like ApplyUpdateEnabled but adds check if the other conditions for IsSupported (see the above thread with Jan about the coreclr version).

lambdageek replied:
@mikem8361 I can fix it up in a follow-up PR, or if you want to take a stab at it, I think you can just modify the implementation of ApplyUpdateEnabled 

https://github.com/dotnet/runtime/blob/0416c3469b7bccff087b89db3d9967dfd36070a3/src/mono/mono/metadata/icall.c#L5790

To something like

```c
return mono_metadata_update_available () && mono_metadata_update_enabled (NULL);
```

(The other callsite of `ApplyUpdateEnabled` is the `GetCapabilities` function - and I think the new version of ApplyUpdateEnabled makes sense there too - we will only return non-empty capabilities if the current running runtime is in a position to apply updates)

mikem8361 replied:
Done

eerhardt commented on src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:
```diff
@@ -12,6 +12,12 @@ public static partial class AssemblyExtensions
+    public static partial class MetadataUpdater
+    {
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta) { throw null; }
+        public static string GetCapabilities() { throw null; }
```
In the API review we said we weren't going to expose `GetCapabilities` publicly yet. Just the other 2.

Do we need to expose it? If so, can you send a mail to fxdr to get it approved?

mikem8361 replied:
Will just an email be sufficient or will I need to create another review issue like #51159?  It doesn't make sense to me for one of the APIs in the new MetadataUpdater class be internal.

eerhardt commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <param name="pdbDelta">The PDB changes to be applied.</param>
+        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta)
```
Do we want to change the existing `AssemblyExtensions` to call this new method, as an interim step before removing it?

mikem8361 replied:
Yes. I'll fix that.

eerhardt commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        /// Returns true if the apply assembly update is enabled and available.
+        /// </summary>
+        public static bool IsSupported => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "" || Environment.GetEnvironmentVariable("COMPlus_ForceEnc") == "1";
```
Do we want to add a linker [feature switch](https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md#adding-new-feature-switch) for this now? If not now, I can add one later if you'd prefer.

mikem8361 replied:
I added the following to https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml and to https://github.com/dotnet/runtime/blob/main/src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml.

I'll add the Microsoft.NET.Sdk.targets msbuild changes in separate sdk PR unless you want to do that :).

```
    <type fullname="System.Reflection.Metadata.MetadataUpdater" feature="System.Reflection.Metadata.MetadataUpdater.IsSupported" featurevalue="false">
      <method signature="System.Boolean get_IsSupported()" body="stub" value="false" />
    </type>
```

lambdageek commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,65 @@
+
+        public static bool IsSupported => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "";
+
+        [MethodImpl (MethodImplOptions.InternalCall)]
+        private static extern int ApplyUpdateEnabled ();
+
+        [MethodImpl (MethodImplOptions.InternalCall)]
```
This needs a corresponding change to src/mono/metadata/icall-def-netcore.h to update the class name:

https://github.com/dotnet/runtime/blob/c0ed319ba3164bfbd9c2c4d66f68f29f8cd78b04/src/mono/mono/metadata/icall-def-netcore.h#L248-L250

Commit:
Code review feedback

Fix the ApplyUpdate qcalls and icalls

Add the ILLink substitutions for MetadataUpdater.IsSupported property

Change the old APIs to call the new ones

Commit:
Update mono's MetadataUpdater.IsSupported property

Commit:
Update feature switch doc

lambdageek commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:
```diff
@@ -82,5 +82,43 @@ void ClassWithCustomAttributes()
+        }
+
+        [Fact]
+        public static void IsSupported()
+        {
+            bool result = MetadataUpdater.IsSupported;
+            Assert.False(result);
+        }
```
I think this test will fail on WebAssembly.  The .csproj sets `DOTNET_MODIFIABLE_ASSEMBLIES=debug` at launch

mikem8361 replied:
What do you suggest for mono? Does the below CI tests include a WebAssembly leg?  It is kind of a silly test maybe I should remove it or skip it on mono.

lambdageek replied:
I think we should skip it on Mono.

I'm not 100% clear on why it didn't fail on the CI leg.  Investigating.  I don't think it's worth holding up the whole PR.

lambdageek reviewed (approved):
Mono bits LGTM.  CoreCLR looks ok, too, but I'm not super-familiar with the code.

One thing to watch for: the `ApplyUpdateTest` testsuite only runs in environments where `ApplyUpdateUtil.IsSupported` is true - for CoreCLR that's basically everywhere, but for Mono in environments where we can't use the remote executor (ie mobile and webassembly) the whole class will only run when `DOTNET_MODIFIABLE_ASSEMBLIES` is already set appropriately by the launcher.

So `MetadataUpdater.IsSupported` will always return true if the AssemblyUpdateTest class is not skipped under Mono.

mikem8361 commented:
All I'm waiting for now is for @eerhardt to approve the feature switch changes and for the API review for making MetadataUpdater.GetCapabilities() public.

eerhardt commented on src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml:
```diff
@@ -6,5 +6,8 @@
     <type fullname="Internal.Runtime.InteropServices.ComponentActivator" feature="System.Runtime.InteropServices.EnableConsumingManagedCodeFromNativeHosting" featurevalue="false">
       <method signature="System.Boolean get_IsSupported()" body="stub" value="false" />
     </type>
+    <type fullname="System.Reflection.Metadata.MetadataUpdater" feature="System.Reflection.Metadata.MetadataUpdater.IsSupported" featurevalue="false">
```
It's a bit unfortunate this is duplicated in the ILLink.Substitutions.xml files between coreclr and mono.

Since this is a public class/property, they won't differ between mono and coreclr, so can you reduce the duplication and put this entry in [https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml) instead?

eerhardt commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,65 @@
+    public static class MetadataUpdater
+    {
+        /// <summary>
+        /// Updates the specified assembly using the provided metadata, IL and PDB deltas.
```
I wonder if we can make this a `partial` class and have all the stuff that is "common" between mono and coreclr shared in `src\libraries\System.Private.CoreLib`? The "public API" could be shared, and then call into the other partial class for the implementation.

Right now the XML doc is duplicated between both.

mikem8361 replied:
The implementations of all 3 APIs are different between coreclr and mono. I don't see anything that would be common exception the XML doc headers.

eerhardt replied:
> The implementations of all 3 APIs are different between coreclr and mono

Maybe forcing things to be more common would lead to better code. I see the argument checking is roughly the same, but different. For example, mono checks if you are trying to apply an update to CoreLib, and throws `InvalidOperationException(SR.InvalidOperation_AssemblyNotEditable)` (which InvalidOperationException is not doc'd BTW). What does coreclr do?

Mono uses the better pattern of `if (assembly is not RuntimeAssembly runtimeAssembly)`.

mikem8361 replied:
Other than argument checking (which I can fix the coreclr version) there isn't anything in common.  Coreclr doesn't need to explicitly check for System.Private.CoreLib because the QCall does that (and it does return InvalidOperationException).  

It really doesn't seem worth to have common code here.

I'll change the coreclr code to use the same argument checking as mono and add InvalidOperationException and ArgumentException to the XML header.

Commit:
Code review feedback. Making the new MetadataUpdater.GetCapabilities() internal.

Fixed the argument checking in coreclr's MetadataUpdater.ApplyUpdate().

eerhardt commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs:
```diff
@@ -0,0 +1,67 @@
+
+        private static string InitializeApplyUpdateCapabilities()
+        {
+            return ApplyUpdateEnabled() != 0 ? "Baseline" : string.Empty ;
```
How come the mono version will return `string.Empty` if ApplyUpdateEnabled returns 0, but the coreclr version always returns the string, and doesn't check if it is supported?

mikem8361 replied:
Because of mono's componentization there are versions of mono that don't support any updating capability (returns empty string).  I guess that there are some more obscure builds of coreclr (Linux x86 I think but that isn't a supported platform) that don't have ENC enabled, but general it is always build in.

eerhardt reviewed (approved):
LGTM

stephentoub commented:
We should also update https://github.com/dotnet/runtime/blob/b1a417875e68e1961c163c1f0611549c2c2e04f6/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml#L34 to be based on the new switch rather than on Debugger.IsSupported.

mikem8361 commented:
How would one make the attribute do that?  The xml file just seems to be the attribute type.

stephentoub commented:
It's currently in the section for the debugger switch:
https://github.com/dotnet/runtime/blob/b1a417875e68e1961c163c1f0611549c2c2e04f6/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml#L3
A new section should be added for the new switch and this attribute moved to that section.

Commit:
Code review feedback.
